airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py

```diff
@@ -335,7 +336,7 @@ def parse_response(
         records = self.record_selector.select_records(
             response=response, stream_state=self.state, stream_slice=stream_slice, next_page_token=next_page_token
         )
-        self._last_records = records
+        self._last_records = copy.deepcopy(records)
```

Comment:
yes I agree, pretty rare and not a very obvious thing

the problem is next:

When we pass record [here-1](https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py#L378) we pass record BEFORE transformer
but when we pass record [here-2](https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py#L383) we pass record AFTER transformer
because `self._last_records = records` keep a reference on mutable record object which is transformed on later stages.

It's a pretty rare case but it affects me on `source-twilio` where I transform `cursor_field` to another datetime format like [this](https://github.com/airbytehq/airbyte/blob/a8f7109a43f0cdfc7ed0127c038d749b08723fd1/airbyte-integrations/connectors/source-twilio/source_twilio/manifest.yaml#L59)

tag: @sherifnada

Reply:
I'm not very familiar with the for/else in Python. My understanding is that "else" is executed if there is no "break" in the loop. I don't see any break statement in the loop above so isn't it [this else clause](https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py#L380) useless because it will be called anyway [as part of the for loop](https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py#L378)?

If we can't remove the else clause, I feel this change makes sense

Reply:
yes, it looks like we can re-write without "else" clause:

```python
for record in records_generator:
    # Only record messages should be parsed to update the cursor which is indicated by the Mapping type
    if isinstance(record, Mapping):
        self.stream_slicer.update_cursor(stream_slice, last_record=record)
    yield record

last_record = self._last_records[-1] if self._last_records else None
if last_record and isinstance(last_record, Mapping):
    self.stream_slicer.update_cursor(stream_slice, last_record=last_record)
yield from []
```

Reply:
My question is "why can we to this?"

```python
for record in records_generator:
    # Only record messages should be parsed to update the cursor which is indicated by the Mapping type
    if isinstance(record, Mapping):
        self.stream_slicer.update_cursor(stream_slice, last_record=record)
    yield record
```

`self.stream_slicer.update_cursor(stream_slice, last_record=last_record)` should be called as part of the for loop and therefore, I'm not sure why we do this last `update_cursor` outside of the loop. Being able to remove this would clean the code and solve the issue you were facing where you had to add `copy.deepcopy(records)`

Reply:
@girarda , based on git blame, I see this is some code you might be familiar with. Can you provide more information as to why this else would be needed?

Reply:
We need the second block to update the state even if a stream slice has no records. 

That being said, we can avoid the duplicate update by only updating with the last record if no records were read

```
read_records = False
for record in records_generator: 
    if isinstance(record, Mapping):
        self.stream_slicer.update_cursor(stream_slice, last_record=record)
        read_records = True
    yield record

if not read_records:
    last_record = self._last_records[-1]
    if self._last_records else None if last_record and isinstance(last_record, Mapping):
        self.stream_slicer.update_cursor(stream_slice, last_record=last_record)
yield from []
```

Reply:
fix is in fix is in https://github.com/airbytehq/airbyte/pull/24827